src/parser/mod.rs

```diff
@@ -331,9 +332,9 @@ pub struct Parser<'a> {
     /// The tokens
     tokens: Vec<TokenWithSpan>,
     /// The index of the first unprocessed token in [`Parser::tokens`].
-    index: usize,
+    index: Cell<usize>,
```

Comment:
This is a clever way to go about reducing the need for mutable borrows. Thank you @eyalleshem 

One thing that might be worth doing (perhaps as a follow on PR) is documenting how this works (notable that Parser has interior mutability so functions like `next_token()` actually do change state, even though they are immutable